### PR TITLE
Preserve functional test logs on CI

### PR DESCRIPTION
**Pull request by JunTaoLuo**
WIP. Testing approach in CI.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -111,11 +111,17 @@ stages:
...
+        - script: eng\common\cibuild.cmd
+            -configuration $(_BuildConfig)
+            -prepareMachine
+            $(_BuildArgs)
+            /bl:artifacts/log/build.windows.binlog
           displayName: Build and Publish
         - script: eng\scripts\ci-flaky-tests.cmd -configuration $(_BuildConfig)
           displayName: Run Flaky Tests
           continueOnError: true
+        - script: dotnet msbuild eng/repo.targets /t:EnsureFunctionalTestLogsPreserved
```
Comment (JunTaoLuo) [inline azure-pipelines.yml]: I suppose I can add the verification to macOS and Linux runs as well, but I don't think it adds much.

File: eng/repo.targets [hunk 2]
```diff
@@ -52,4 +52,16 @@
           SkipNonexistentTargets="true"
           SkipNonexistentProjects="true" />
   </Target>
+
+  <Target Name="EnsureFunctionalTestLogsPreserved">
+    <PropertyGroup>
+      <FunctionalLogDir>$(ArtifactsDir)log\Microsoft.Extensions.Logging.Testing.Tests\$(DefaultNetCoreTargetFramework)\AssemblyTestLogTests\</FunctionalLogDir>
+      <FlakyTestLog>$(FunctionalLogDir)FunctionalLogs_LogsPreservedFromFlakyRun.log</FlakyTestLog>
```
Comment (JunTaoLuo) [inline eng/repo.targets]: The only complaint I have is that this is fragile against renames, but oh well.

File: src/Logging/Logging.Testing/src/build/Microsoft.Extensions.Logging.Testing.props [hunk 3]
```diff
@@ -9,15 +9,21 @@
   <Target Name="SetLoggingTestingAssemblyAttributes"
     BeforeTargets="GetAssemblyAttributes"
     Condition="'$(GenerateLoggingTestingAssemblyAttributes)' != 'false'">
+    <PropertyGroup>
+      <PreserveExistingLogsInOutput Condition="'$(PreserveExistingLogsInOutput)' == '' AND '$(ContinuousIntegrationBuild)' == 'true'">true</PreserveExistingLogsInOutput>
```
Comment (JunTaoLuo) [inline src/Logging/Logging.Testing/src/build/Microsoft.Extensions.Logging.Testing.props]: The idea is that we will preserve existing logs on CI but delete existing logs locally by default. These can always be overridden if needed.

File: azure-pipelines.yml (hunk 1 above)
Comment (BrennanConroy) [reply azure-pipelines.yml]: Meh

File: src/TestingUtils/Microsoft.AspNetCore.Testing/ref/Microsoft.AspNetCore.Testing.net46.cs [hunk 4]
```diff
@@ -295,6 +295,7 @@ public sealed partial class TestFileOutputContext
         public string TestName { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } }
         public static string GetAssemblyBaseDirectory(System.Reflection.Assembly assembly, string baseDirectory = null) { throw null; }
         public static string GetOutputDirectory(System.Reflection.Assembly assembly) { throw null; }
+        public static bool GetPreserveExistingLogsInOutput(System.Reflection.Assembly assembly) { throw null; }
```
Comment (dougbu) [inline src/TestingUtils/Microsoft.AspNetCore.Testing/ref/Microsoft.AspNetCore.Testing.net46.cs]: You won't need to do this if the ref/ assembly removal happens first. But, I guess that has to wait for branches to open ☹️

Comment (JunTaoLuo) [reply src/TestingUtils/Microsoft.AspNetCore.Testing/ref/Microsoft.AspNetCore.Testing.net46.cs]: Yup

**Pull request merged**